Ekubo_wstETH_ETH_strategy.ipynb

```diff
@@ -0,0 +1,1225 @@
+    "optimization_results = []\n",
+    "block_vs_returns_all = []\n",
+    "\n",
+    "# Use itertools.product to generate combinations of liquidity and rebalance ranges\n",
+    "for liquidity_range, rebalance_range in itertools.product(liquidity_range_options, rebalance_range_options):\n",
+    "    # Run the simulation for each combination\n",
+    "    block_vs_returns_df, net_earnings, return_percentage, fee_earned_eth, impermanent_loss_total, current_liquidity = strategy.run_simulation(\n",
+    "        liquidity_range,\n",
+    "        rebalance_range,\n",
+    "        result['train_swap_df'],  \n",
+    "        ETH_provided=ETH_provided, \n",
+    "        fee_percentage=fee_percent_fraction, \n",
+    "        decimal=1  \n",
+    "    )\n",
+    "    \n",
+    "    # Append results to the optimization list\n",
+    "    optimization_results.append({\n",
+    "        \"Liquidity Range (%)\": liquidity_range * 100,\n",
+    "        \"Rebalance Range (%)\": rebalance_range * 100,\n",
+    "        \"Returns (%)\": return_percentage,\n",
+    "        \"Net Earnings (ETH)\": net_earnings,\n",
+    "        \"Total Fees Earned (ETH)\": fee_earned_eth,\n",
+    "        \"Impermanent Loss (ETH)\": impermanent_loss_total,\n",
+    "    })\n",
+    "    \n",
+    "    # Store the block vs. returns DataFrame for plotting\n",
+    "    block_vs_returns_all.append(block_vs_returns_df)\n",
+    "\n",
+    "# Convert the optimization results to a DataFrame for analysis\n",
+    "optimization_df = pd.DataFrame(optimization_results)\n",
+    "\n",
+    "# Sort by net earnings\n",
+    "optimization_df_sorted = optimization_df.sort_values(by=\"Net Earnings (ETH)\", ascending=False)\n",
+    "\n",
+    "# Display the top 5 performing combinations\n",
+    "top_results = optimization_df_sorted.head()\n",
+    "print(top_results)\n",
+    "\n",
+    "# Plotting function to visualize returns vs. block number (time)\n",
+    "def plot_returns_vs_block(returns_df, title=\"Returns vs Block Number\"):\n",
+    "    \"\"\"\n",
+    "    Plot the returns vs. block number curve.\n",
+    "\n",
+    "    Parameters:\n",
+    "    - returns_df (pd.DataFrame): DataFrame containing 'block_number' and 'returns' columns.\n",
+    "    - title (str): Title of the plot.\n",
+    "    \"\"\"\n",
+    "    plt.figure(figsize=(10, 6))\n",
+    "    plt.plot(returns_df['block_number'], returns_df['returns'], label=\"Returns\", color='blue')\n",
+    "    plt.xlabel('Block Number')\n",
+    "    plt.ylabel('Returns (%)')\n",
+    "    plt.title(title)\n",
+    "    plt.legend()\n",
+    "    plt.grid(True)\n",
+    "    plt.show()\n",
+    "\n",
+    "# Assuming you want to plot the returns for the best performing combination\n",
+    "best_block_vs_returns_df = block_vs_returns_all[optimization_df_sorted.index[0]]  # Get the best block vs returns\n",
+    "plot_returns_vs_block(best_block_vs_returns_df, title=\"Returns vs Block Number for Best Strategy\")\n"
```

Comment:
In this plot, add vertical lines at which blocks rebalance happen. 
Second, also plot the block wise lower and upper price used to add liquidity. 
plot fees earned and IL as well wrt to block. 

Do this plot for 0.1% liquidity range, plus the best strategy as u r doing now.